### DISCUSSION TITLE
How to code the following constraint to make sure it follows DCP rules?

### DISCUSSION THREAD

**Discussion by s0ap** [cvxpy/cvxpy#1759] [Q&A] [answered]
I don't have much experience with cvxpy atomic expressions but could someone help me?

I’ve been trying to implement a constraint to minimise tax liability on a ETF (using let's say semi annual rebalance schedule). However, the objective I've coded seems to be non-convex (is not DCP compliant).

I have a trade data frame, with a union of all tickers from previous and current rebalance as my index - along with number of shares and price at each date. Please see below for a sample of the data i've been working with (init_weights are benchmark weights) and my interpretation of how to add the tax liability constraint. 

<img width="469" alt="Screenshot 2022-04-27 at 09 45 15" src="https://user-images.githubusercontent.com/13176243/165533041-ec586520-7317-4730-b97b-088ccab8f66e.png">

```python
import cvxpy as cp
n = len(trade_df.index)
# New shares to solve for
S_t = cp.Variable(n)

tax_liability = cp.sum(cp.multiply(cp.neg(S_t - trade_df['Prev Shares'].values), trade_df['price_diff'].values))

w = cp.multiply(S_t, trade_df['Curr Price'].values) / 
                           cp.sum(cp.multiply(trade_df['Curr Price'].values, trade_df['Curr Shares'].values))
rel_weights = w - trade_df['init_weights'].values
tracking_error = cp.quad_form(rel_weights, S)

objective = cp.Minimize(tax_liability)
constraints = [cp.sum(w) == 1, w >= 0]
prob = cp.Problem(objective, constraints)

prob.solve(verbose=True)
```
The error is "Problem does not follow DCP rules. Specifically: The objective is not DCP." 

I think this is due to the fact that price_diff (Curr Price - Prev Price) which I multiply with the number of shares sold, can be positive or negative which renders the problem non-convex.

Can anyone help here? Many thanks!

**Comment by phschiele** [comment, marked as answer]
Hi @s0ap,

cool application!

Indeed, the sign of the price difference makes the objective non-dcp:

```py
import cvxpy as cp
n = 10
# New shares to solve for
S_t = cp.Variable(n)
prev_shares = np.random.random(n)
price_diff = np.random.random(n) # between 0 and 1

tax_liability = cp.sum(cp.multiply(cp.neg(S_t - prev_shares), price_diff))
tax_liability.is_dcp()
>>> True

tax_liability = cp.sum(cp.multiply(cp.neg(S_t - prev_shares), price_diff - 0.5))
tax_liability.is_dcp()
>>> False
```
You can play around with the [DCP analyzer](https://dcp.stanford.edu/analyzer?expr=sum(pos(-(x%20-%201))%20*%202)) to see this in detail. (Note: I had to replace `neg(x)` with `pos(-x)`)

There are a few ways around this issue:
1. Specify beforehand which securities to buy and which to sell
2. Try to replace `S_t` with `S_current + buy_shares - sell_shares` (both of which are **nonnegative** variables of length `n`)

Under 2. you might face the problem that under some constraints the implicit assumption of `buy_shares * sell_shares = 0` could be violated. In this case, you might need to include binary helper variables. 

An interesting reference for Tax-Aware Portfolio Construction via Convex Optimization:
https://arxiv.org/pdf/2008.04985.pdf

**Reply by s0ap** [reply]
Thank you for looking into it! I'm looking at the same paper :) — it's just that they use dollar value of buying/selling shares (<img src="https://render.githubusercontent.com/render/math?math=u">) for formulating the problem (as per my understanding) while I'm trying to solve for number of shares to buy/sell.

I can implement 1. quite easily — since I'm going to track the benchmark. How would you suggest adding this particular constraint?

Thanks!

**Reply by phschiele** [reply]
I would suggest creating two slices from the original dataframe:
One for the overweight positions (having length `n_1`) and one for the underweight positions (having length `n_2`).
Then you define something like:
```py
sell_shares = cp.Variable(n_1, nonneg=True)
S_t_sell = S_current - sell_shares

buy_shares = cp.Variable(n_2, nonneg=True)
S_t_buy = S_current + buy_shares

tax_liability = cp.sum(cp.multiply(sell_shares, price_diff_1))

S_t = cp.hstack([S_t_sell, S_t_buy]) # Use this for the tracking error constraint
```

P.S.: In this part of your original snippet you dont need cvxpy functions, you can use numpy, as there are only constants invovled:
`cp.sum(cp.multiply(trade_df['Curr Price'].values, trade_df['Curr Shares'].values))`
